src/modules/musig/example.c

```diff
@@ -0,0 +1,145 @@
+int sign(const secp256k1_context* ctx, unsigned char seckeys[][32], const secp256k1_pubkey* pubkeys, const unsigned char* msg32, secp256k1_schnorrsig *sig) {
+    secp256k1_musig_session musig_session[3];
+    unsigned char nonce_commitment[N_SIGNERS][32];
+    const unsigned char *nonce_commitment_ptr[N_SIGNERS];
+    secp256k1_musig_session_signer_data signer_data[N_SIGNERS][N_SIGNERS];
+    secp256k1_pubkey nonce[N_SIGNERS];
+    int i, j;
+    secp256k1_musig_partial_signature partial_sig[N_SIGNERS];
+
+    for (i = 0; i < N_SIGNERS; i++) {
+        FILE *frand;
+        unsigned char session_id32[32];
+        unsigned char pk_hash[32];
+        secp256k1_pubkey combined_pk;
+
+        /* Create combined pubkey and initialize signer data */
+        if (!secp256k1_musig_pubkey_combine(ctx, NULL, &combined_pk, pk_hash, pubkeys, N_SIGNERS)) {
+            return 0;
+        }
+        /* Create random session ID. It is absolutely necessary that the session ID
+         * is unique for every call of secp256k1_musig_session_initialize. */
+        frand = fopen("/dev/urandom", "r");
+        if (frand == NULL || !fread(session_id32, 32, 1, frand)) {
+            return 0;
+        }
+        fclose(frand);
+        /* Initialize session */
+        if (!secp256k1_musig_session_initialize(ctx, &musig_session[i], signer_data[i], nonce_commitment[i], session_id32, msg32, &combined_pk, pk_hash, pubkeys, N_SIGNERS, i, seckeys[i])) {
+            return 0;
+        }
+        nonce_commitment_ptr[i] = nonce_commitment[i];
+    }
+    /* Communication round 1: Exchange nonce commitments */
+    for (i = 0; i < N_SIGNERS; i++) {
+        /* Set nonce commitments in the signer data and get the own public nonce */
+        if (!secp256k1_musig_session_get_public_nonce(ctx, &musig_session[i], signer_data[i], &nonce[i], nonce_commitment_ptr, N_SIGNERS)) {
+            return 0;
+        }
+    }
+    /* Communication round 2: Exchange nonces */
+    for (i = 0; i < N_SIGNERS; i++) {
+        for (j = 0; j < N_SIGNERS; j++) {
+            if (!secp256k1_musig_set_nonce(ctx, &signer_data[i][j], &nonce[j])) {
+                /* Signer j's nonce does not match the nonce commitment. Wait
+                 * until the correct nonce is received or restart the protocol
+                 * (with a different session ID of course). */
+                return 0;
+            }
+        }
+        if (!secp256k1_musig_session_combine_nonces(ctx, &musig_session[i], signer_data[i], N_SIGNERS, NULL, NULL)) {
+            return 0;
+        }
+    }
+    for (i = 0; i < N_SIGNERS; i++) {
+        if (!secp256k1_musig_partial_sign(ctx, &musig_session[i], &partial_sig[i])) {
+            return 0;
+        }
+    }
+    /* Communication round 3: Exchange partial signatures */
+    for (i = 0; i < N_SIGNERS; i++) {
+        for (j = 0; j < N_SIGNERS; j++) {
+            if (!secp256k1_musig_partial_sig_verify(ctx, &musig_session[i], &signer_data[i][j], &partial_sig[j])) {
```

Comment:
Maybe add a note on the need/non-need for partial verification.

Reply:
Do you have a suggestion? In practice I think you'll always want to verify partial sigs if just for debugging.

Reply:
"To check whether signing was successful, it suffices to either verify the the combined signature with the combined public key, or verify all partial signatures of all signers individually. Verifying the combined signature is cheaper but verifying the individual signatures has the advantage that it can be used to determine which of the signers misbehaved, i.e., sent invalid partial signatures in the protocol. " or something

I think in practice, it's also good to first verify the combined sig, and only verify the individual sigs if it does not work.

Reply:
fixing

Reply:
nit: sorry, the text was my suggestion but now I realize that "misbehaved" is too much. There could be a network attacker flipping bits etc.
"Verifying the combined signature is cheaper but verifying the individual partial signatures has the advantage that it can be used to determine which of the partial signatures are invalid (if any), i.e., which of the partial signatures cause the combined signature to be invalid and thus the protocol run to fail."

Reply:
I think "misbehaved" is fine, though it does have a bit of a connotation that there was deliberate action.